src/common.c

```diff
@@ -1196,22 +1196,27 @@ int ofi_bsock_flush_sync(struct ofi_bsock *bsock)
 	return ofi_bsock_tosend(bsock) ? -FI_EAGAIN : 0;
 }
 
-ssize_t ofi_bsock_send(struct ofi_bsock *bsock, const void *buf, size_t *len)
+int ofi_bsock_send(struct ofi_bsock *bsock, const void *buf, size_t *len)
 {
 	size_t avail;
 	ssize_t ret;
+	int err;
 
 	avail = ofi_bsock_tosend(bsock);
 	if (avail) {
 		if (*len < ofi_byteq_writeable(&bsock->sq)) {
 			ofi_byteq_write(&bsock->sq, buf, *len);
-			ret = ofi_bsock_flush(bsock);
-			return !ret || ret == -FI_EAGAIN ? *len : ret;
+			err = ofi_bsock_flush(bsock);
+			if (err && err != -FI_EAGAIN)
+				*len = 0;
+			return err;
 		}
 
-		ret = ofi_bsock_flush(bsock);
-		if (ret)
-			return ret;
+		err = ofi_bsock_flush(bsock);
+		if (err) {
+			*len = 0;
+			return err;
+		}
```

Comment:
In the commit message, it mentions 'bytes received' rather than 'bytes sent'.  But from the code, I can't tell what len is supposed to be set to.  In the above flows, len is set to 0, but some data was actually queued on the bsock buffer.  Below, len is set to the number of bytes sent on the socket.

Reply:
I see what you mean. IMO, *len should always contain the number of bytes consumed by the bsock interface for the given call, whether the data is buffered in the bsock buffer or sent on the socket. 
In the case where `ofi_bsock_flush()` fails with an error different that -FI_EAGAIN, I would be tempted to rollback the changes to the bsock buffer (i.e., `byteq->tail -= *len`) and set `*len = 0`. If `ofi_bsock_flush()` returns `0` or -FI_EAGAIN, we cab consider that the function ofi_bsock_send() was successful () and return 0.

For example:
```
avail = ofi_bsock_tosend(bsock);                                    
if (avail) {                                                        
        if (*len < ofi_byteq_writeable(&bsock->sq)) {               
                ofi_byteq_write(&bsock->sq, buf, *len);             
                err = ofi_bsock_flush(bsock);                       
                if (err && err != -FI_EAGAIN) {                     
                        /* Add a call to rollback the changes done by ofi_byteq_write(): byteq->tail -= *len */
                        *len = 0;                                   
                        return err;                                 
                }                                                   
                return 0;                                           
        }                                                           
                                                                    
        err = ofi_bsock_flush(bsock);                               
        if (err) {                                                  
                *len = 0;                                           
                return err;                                         
        }                                                           
}                                                                   
```

But it seems to me that the problem existed because my change, no? Say if `ofi_bsock_flush()` fails with  an error different than -FI_EAGAIN. In this case the data is queued to the bsock buffer, but the caller doesn't know about it.

Reply:
If flush fails with anything other the EAGAIN, the socket is in an error state, and there's no telling what may have made it to the other side.  I would not try to rollback or anything like that.  The transfer is doomed anyway.  I wouldn't even bother adjusting len based on that, though it should be harmless to set it to 0.

Reply:
I updated the code with the following changes to ofi_bsock_send/ofi_bsock_sendv: 
- I clarified the usage of "*len" in the commit message.
- I fixed the error handling of ofi_bsock_flush. Here is how the code looks now:
```
if (*len < ofi_byteq_writeable(&bsock->sq)) {
        ofi_byteq_writev(&bsock->sq, iov, cnt);
        err = ofi_bsock_flush(bsock);
        return !err || err == -FI_EAGAIN ? 0 : err;
}
```